### PR TITLE
Updated objfw

### PR DESCRIPTION
**Pull request by Frityet**

### CODE REVIEW

**Comment by Frityet** [comment]
Don't merge yet please! Give me just one moment

**Comment by Frityet** [comment]
Ready for review

**Comment by Frityet** [comment]
fixing tests...

**Comment by Frityet** [comment]
Ready :+1:

File: packages/o/objfw/xmake.lua [hunk 1]
```diff
@@ -12,13 +12,94 @@ package("objfw")
...
+        config("codepage-858")
+        config("codepage-850")
+        config("iso-8859-3")
+        config("iso-8859-2")
+        config("koi8-r")
+        config("koi8-u")
+        config("iso-8859-15")
+        config("mac-roman")
+        config("windows-1251")
+        config("windows-1252")
```
Comment (waruqi) [inline packages/o/objfw/xmake.lua:78]: you can for-each package:configs()

see

https://github.com/xmake-io/xmake-repo/blob/2f0b2cb3c771dd81121d05bc8df68516c749ff87/packages/f/ffmpeg/xmake.lua#L122

File: packages/o/objfw/xmake.lua [hunk 2]
```diff
@@ -12,13 +12,94 @@ package("objfw")
...
+
+        if package:is_debug() then
             table.insert(configs, "--enable-debug")
         end
         import("package.tools.autoconf").install(package, configs)
+
+        local objfwcfg = import("lib.detect.find_tool")("objfw-config", { paths = { package:installdir("bin") } }).program
```
Comment (waruqi) [inline packages/o/objfw/xmake.lua:85]:
```
local objfwcfg = path.join(package:installdir("bin"), "objfw-config")
```

File: packages/o/objfw/xmake.lua [hunk 3]
```diff
@@ -12,13 +12,94 @@ package("objfw")
...
+
+        local objcflags_str = os.iorunv(objfwcfg, { "--objcflags", (package:config("arc") and "--arc" or "") })
+        local ldflags_str = os.iorunv(objfwcfg, { "--ldflags" })
+
+        local objcflags = {}
+        local ldflags = {}
+
+        for flag in objcflags_str:gmatch("%S+") do
```
Comment (waruqi) [inline packages/o/objfw/xmake.lua:93]:
```
objcflags_str:split("%s")
```

File: packages/o/objfw/xmake.lua [hunk 4]
```diff
@@ -12,13 +12,94 @@ package("objfw")
...
+            table.insert(objcflags, flag)
+        end
+
+        for flag in ldflags_str:gmatch("%S+") do
```
Comment (waruqi) [inline packages/o/objfw/xmake.lua:97]: and here

File: packages/o/objfw/xmake.lua [hunk 5]
```diff
@@ -12,13 +12,94 @@ package("objfw")
...
         import("package.tools.autoconf").install(package, configs)
+
+        local objfwcfg = import("lib.detect.find_tool")("objfw-config", { paths = { package:installdir("bin") } }).program
+
+        local objcflags_str = os.iorunv(objfwcfg, { "--objcflags", (package:config("arc") and "--arc" or "") })
+        local ldflags_str = os.iorunv(objfwcfg, { "--ldflags" })
+
+        local objcflags = {}
+        local ldflags = {}
+
```
Comment (waruqi) [inline packages/o/objfw/xmake.lua:92]: remove empty line

File: packages/o/objfw/xmake.lua [hunk 6]
```diff
@@ -12,13 +12,94 @@ package("objfw")
         add_frameworks("CoreFoundation")
     end
 
-    on_install("linux", "macosx", function (package)
-        local configs = {"--without-tls"}
+    add_configs("tls", { description = "Enable TLS support.", default = true, values = { true, false, "yes", "no", "openssl", "gnutls", "securetransport" } })
```
Comment (waruqi) [inline packages/o/objfw/xmake.lua:15]: remove yes/no, only use true/false

File: packages/o/objfw/xmake.lua [hunk 7]
```diff
@@ -12,13 +12,94 @@ package("objfw")
...
+    add_configs("files", { description = "Enable files.", default = true, type = "boolean" })
+    add_configs("sockets", { description = "Enable sockets.", default = true, type = "boolean" })
+
+    add_configs("arc", { description = "Enable Automatic Reference Counting (ARC) support.", default = true, type = "boolean" })
+
+    on_install("linux", "macosx", "cygwin", "mingw", function (package)
+        local configs = {}
+
+        local tls = package:config("tls")
+        tls = tls == true and "yes" or tls == false and "no" or tls
```
Comment (waruqi) [inline packages/o/objfw/xmake.lua:45]:
```
if type(tls) == "boolean" then
else
end
```

**Comment by Frityet** [comment]
@waruqi is this ready to be merged?

**Comment by waruqi** [comment]
> @waruqi is this ready to be merged?

why? some ci tests are failure

**Comment by Frityet** [comment]
> > @waruqi is this ready to be merged?
> 
> why? some ci tests are failure

A few of them I don't really know how to fix, such as

```
 note: the following packages are unsupported on mingw/x86_64:
  -> autoconf 2.71
  -> automake 1.16.4
  -> libtool 2.4.6
error: @programdir\core\main.lua:300: stack traceback:
    [C]: in function 'error'
    [@programdir\core\base\os.lua:924]:
    [...modules\private\action\require\impl\install_packages.lua:744]:
    [@programdir\modules\private\action\require\install.lua:85]:
    [@programdir\actions\require\main.lua:123]:
    [C]: in function 'xpcall'
    [@programdir\core\base\utils.lua:280]:
    [@programdir\core\base\task.lua:501]: in function 'run'
    [@programdir\core\main.lua:298]: in function 'cotask'
    [@programdir\core\base\scheduler.lua:404]:

stack traceback:
	[C]: in function 'error'
	@programdir\core\base\os.lua:913: in function 'os.raiselevel'
	(...tail calls...)
	@programdir\core\main.lua:300: in upvalue 'cotask'
	@programdir\core\base\scheduler.lua:404: in function <@programdir\core\base\scheduler.lua:397>
error: execv(xmake require -f -y --build -v -D --extra={configs={shared=true}} objfw) failed(-1)
Error: Process completed with exit code 127.
```

Even though msys2 should be able to run `configure` and `autoconf`

**Comment by waruqi** [comment]
> > > @waruqi is this ready to be merged?
> > 
> > 
> > why? some ci tests are failure
> 
> A few of them I don't really know how to fix, such as
> 
> ```
>  note: the following packages are unsupported on mingw/x86_64:
>   -> autoconf 2.71
>   -> automake 1.16.4
>   -> libtool 2.4.6
> error: @programdir\core\main.lua:300: stack traceback:
>     [C]: in function 'error'
>     [@programdir\core\base\os.lua:924]:
>     [...modules\private\action\require\impl\install_packages.lua:744]:
>     [@programdir\modules\private\action\require\install.lua:85]:
>     [@programdir\actions\require\main.lua:123]:
>     [C]: in function 'xpcall'
>     [@programdir\core\base\utils.lua:280]:
>     [@programdir\core\base\task.lua:501]: in function 'run'
>     [@programdir\core\main.lua:298]: in function 'cotask'
>     [@programdir\core\base\scheduler.lua:404]:
> 
> stack traceback:
> 	[C]: in function 'error'
> 	@programdir\core\base\os.lua:913: in function 'os.raiselevel'
> 	(...tail calls...)
> 	@programdir\core\main.lua:300: in upvalue 'cotask'
> 	@programdir\core\base\scheduler.lua:404: in function <@programdir\core\base\scheduler.lua:397>
> error: execv(xmake require -f -y --build -v -D --extra={configs={shared=true}} objfw) failed(-1)
> Error: Process completed with exit code 127.
> ```
> 
> Even though msys2 should be able to run `configure` and `autoconf`

These packages do not yet support msys2. You'll need to mention pr to enable them and fix the problems encountered. Alternatively, you can disable the mingw platform for the current package.

**Comment by waruqi** [comment]
mingw/gcc does not support objc

File: packages/o/objfw/xmake.lua [hunk 8]
```diff
@@ -1,24 +1,104 @@
...
+            table.insert(objcxxflags, flag)
+        end
+        for _, flag in ipairs(ldflags_str:split("%s+")) do
+            table.insert(ldflags, flag)
+        end
+        package:add("mflags", objcflags)
+
+        --does not work?
+        package:add("mxxflags", objcxxflags)
+        package:add("ldflags", ldflags)
```
Comment (waruqi) [inline packages/o/objfw/xmake.lua:101]: ?

Comment (waruqi) [reply packages/o/objfw/xmake.lua:93]: I have fixed it for mxxflags in xmake/dev

**Comment by waruqi** [comment]
> https://objfw.nil.im/doc/trunk/README.md#windows

maybe it need mingw/clang, but msys2/mingw and macos/mingw is gcc on ci

**Comment by Frityet** [comment]
> 

hmmm, is it possible to make it require clang on windows?

**Pull request merged**